packages/hasher/src/FileHasher.ts

```diff
@@ -0,0 +1,121 @@
+import fs from "graceful-fs";
+import path from "path";
+import { hash as fastHash, stat } from "glob-hasher";
+import { createInterface } from "node:readline";
+import { getPackageDeps } from "./getPackageDeps.js";
+
+interface FileHashStoreOptions {
+  root: string;
+}
+
+type FileHashManifestStore = Record<
+  string,
+  {
+    mtime: bigint;
+    size: number;
+    hash: string;
+  }
+>;
+
+export class FileHasher {
+  #store: FileHashManifestStore = {};
+  #manifestFile: string;
+
+  constructor(private options: FileHashStoreOptions) {
+    const { root } = options;
+    const cacheDirectory = path.join(root, "node_modules", ".cache", "lage");
+    this.#manifestFile = path.join(cacheDirectory, "file_hashes.manifest");
+  }
+
+  async getHashesFromGit() {
+    const { root } = this.options;
+    const fileHashes = await getPackageDeps(root);
+    const files = [...fileHashes.keys()];
+    const fileStats = stat(files, { cwd: root }) ?? {};
+
+    for (const [relativePath, fileStat] of Object.entries(fileStats)) {
+      if (fileHashes.has(relativePath)) {
+        this.#store[relativePath] = {
+          mtime: fileStat.mtime,
+          size: fileStat.size,
+          hash: fileHashes.get(relativePath)!,
+        };
+      }
+    }
+
+    this.writeManifest();
+  }
+
+  async readManifest() {
+    return new Promise<void>((resolve) => {
+      if (!fs.existsSync(this.#manifestFile)) {
+        this.getHashesFromGit().then(() => resolve());
+        return;
+      }
+
+      const inputStream = fs.createReadStream(this.#manifestFile, "utf-8");
+      const rl = createInterface({
+        input: inputStream,
+        crlfDelay: Infinity,
+      });
+
+      rl.on("line", (line) => {
+        const [relativePath, mtimeStr, sizeStr, hash] = line.split("\0");
+        this.#store[relativePath] = {
+          mtime: BigInt(mtimeStr),
+          size: parseInt(sizeStr),
+          hash,
+        };
+      });
+
+      inputStream.on("end", () => {
+        rl.close();
+        resolve();
+      });
+    });
+  }
+
+  writeManifest() {
```

Comment:
kinda odd that read is async but write is sync. (not a big deal, just was odd)

Reply:
Yeah, streams are weird for the read -> readline interface thing - needed to convert it to async. Writing in sync basically blocks me from doing too many async writes to that file. I am also using graceful-fs to make sure our watches would be okay if lots of concurrent requests came in. On another PR, I'd probably convert this to an async with a debouncer.